flake: enterprise/wsproxy TestDERPMesh/Proxy9: packet came from wrong source

Seen here: https://github.com/coder/coder/actions/runs/12810577955/job/35717978750#step:8:629

```
=== FAIL: enterprise/wsproxy TestDERPMesh/Proxy9 (0.01s)
    wsproxy_test.go:548: derp1=http://localhost:59979/derp, derp2=http://localhost:60028/derp
    wsproxy_test.go:1088: client1: derphttp.Client.Connect: connecting to http://localhost:59979/derp
    wsproxy_test.go:1088: client1: derphttp.Client.Connect: not using fast start
    wsproxy_test.go:1088: client1: derphttp.Client.Connect: DERP server returned status 101
    wsproxy_test.go:1088: client2: derphttp.Client.Connect: connecting to http://localhost:60028/derp
    wsproxy_test.go:1103: Recv error: derp.Recv: read tcp 127.0.0.1:61087->127.0.0.1:59979: wsarecv: An existing connection was forcibly closed by the remote host.
    wsproxy_test.go:1088: client2: derphttp.Client.Connect: not using fast start
    wsproxy_test.go:1088: client2: derphttp.Client.Connect: DERP server returned status 101
    wsproxy_test.go:1088: client1: derphttp.Client.Send: connecting to http://localhost:59979/derp
    wsproxy_test.go:1088: client1: derphttp.Client.Send: not using fast start
    wsproxy_test.go:1088: client1: derphttp.Client.Send: DERP server returned status 101
    wsproxy_test.go:557: 
        	Error Trace:	C:/a/coder/coder/enterprise/wsproxy/wsproxy_test.go:1144
        	            				C:/a/coder/coder/enterprise/wsproxy/wsproxy_test.go:557
        	Error:      	Not equal: 
        	            	expected: key.NodePublic{k:[32]uint8{0xd2, 0x1b, 0x48, 0x3e, 0xed, 0x75, 0x18, 0x5c, 0xd3, 0x68, 0x46, 0xb6, 0x72, 0x75, 0x7c, 0xa, 0xca, 0xd1, 0xd3, 0xad, 0xaa, 0xf2, 0x2b, 0xde, 0x43, 0x7e, 0xc3, 0x27, 0x56, 0x6, 0x3, 0x1b}}
        	            	actual  : key.NodePublic{k:[32]uint8{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,4 +2,4 @@
        	            	  k: ([32]uint8) (len=32) {
        	            	-  00000000  d2 1b 48 3e ed 75 18 5c  d3 68 46 b6 72 75 7c 0a  |..H>.u.\.hF.ru|.|
        	            	-  00000010  ca d1 d3 ad aa f2 2b de  43 7e c3 27 56 06 03 1b  |......+.C~.'V...|
        	            	+  00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
        	            	+  00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
        	            	  }
        	Test:       	TestDERPMesh/Proxy9
        	Messages:   	packet came from wrong source
    --- FAIL: TestDERPMesh/Proxy9 (0.01s)
```

--- a/enterprise/wsproxy/wsproxy_test.go
+++ b/enterprise/wsproxy/wsproxy_test.go
@@ -838,27 +838,33 @@ func TestWorkspaceProxyDERPMeshProbe(t *testing.T) {
 		require.NoError(t, err)
 
 		// Create 1 real proxy replica.
-		replicaPingErr := make(chan string, 4)
+		replicaPingRes := make(chan replicaPingCallback, 4)
 		proxy := coderdenttest.NewWorkspaceProxyReplica(t, api, client, &coderdenttest.ProxyOptions{
 			Name:     "proxy-2",
 			ProxyURL: proxyURL,
-			ReplicaPingCallback: func(_ []codersdk.Replica, err string) {
-				replicaPingErr <- err
+			ReplicaPingCallback: func(replicas []codersdk.Replica, err string) {
+				t.Logf("got wsproxy ping callback: replica count: %v, ping error: %s", len(replicas), err)
+				replicaPingRes <- replicaPingCallback{
+					replicas: replicas,
+					err:      err,
+				}
 			},
 		})
 
+		// Create a second proxy replica that isn't working.
 		ctx := testutil.Context(t, testutil.WaitLong)
 		otherReplicaID := registerBrokenProxy(ctx, t, api.AccessURL, proxyURL.String(), proxy.Options.ProxySessionToken)
 
-		// Force the proxy to re-register immediately.
-		err = proxy.RegisterNow()
-		require.NoError(t, err, "failed to force proxy to re-register")
-
-		// Wait for the ping to fail.
+		// Force the proxy to re-register and wait for the ping to fail.
 		for {
-			replicaErr := testutil.TryReceive(ctx, t, replicaPingErr)
-			t.Log("replica ping error:", replicaErr)
-			if replicaErr != "" {
+			err = proxy.RegisterNow()
+			require.NoError(t, err, "failed to force proxy to re-register")
+
+			pingRes := testutil.TryReceive(ctx, t, replicaPingRes)
+			// We want to ensure that we know about the other replica, and the
+			// ping failed.
+			if len(pingRes.replicas) == 1 && pingRes.err != "" {
+				t.Log("got failed ping callback for other replica, continuing")
 				break
 			}
 		}
@@ -884,17 +890,17 @@ func TestWorkspaceProxyDERPMeshProbe(t *testing.T) {
 		})
 		require.NoError(t, err)
 
-		// Force the proxy to re-register immediately.
-		err = proxy.RegisterNow()
-		require.NoError(t, err, "failed to force proxy to re-register")
-
-		// Wait for the ping to be skipped.
+		// Force the proxy to re-register and wait for the ping to be skipped
+		// because there are no more siblings.
 		for {
-			replicaErr := testutil.TryReceive(ctx, t, replicaPingErr)
-			t.Log("replica ping error:", replicaErr)
+			err = proxy.RegisterNow()
+			require.NoError(t, err, "failed to force proxy to re-register")
+
+			replicaErr := testutil.TryReceive(ctx, t, replicaPingRes)
 			// Should be empty because there are no more peers. This was where
 			// the regression was.
-			if replicaErr == "" {
+			if len(replicaErr.replicas) == 0 && replicaErr.err == "" {
+				t.Log("got empty ping callback with no sibling replicas, continuing")
 				break
 			}
 		}
@@ -991,6 +997,11 @@ func TestWorkspaceProxyWorkspaceApps(t *testing.T) {
 			FlushStats:     flushStats,
 		}
 	})
+}
+
+type replicaPingCallback struct {
+	replicas []codersdk.Replica
+	err      string
 }
 
 func TestWorkspaceProxyWorkspaceApps_BlockDirect(t *testing.T) {

--- a/enterprise/wsproxy/wsproxy_test.go
+++ b/enterprise/wsproxy/wsproxy_test.go
@@ -498,8 +498,8 @@ func TestDERPMesh(t *testing.T) {
 	proxyURL, err := url.Parse("https://proxy.test.coder.com")
 	require.NoError(t, err)
 
-	// Create 6 proxy replicas.
-	const count = 6
+	// Create 3 proxy replicas.
+	const count = 3
 	var (
 		sessionToken = ""
 		proxies      = [count]coderdenttest.WorkspaceProxy{}

--- a/enterprise/wsproxy/wsproxy_test.go
+++ b/enterprise/wsproxy/wsproxy_test.go
@@ -1,6 +1,7 @@
 package wsproxy_test
 
 import (
+	"bytes"
 	"context"
 	"encoding/json"
 	"fmt"
@@ -1129,7 +1130,7 @@ func createDERPClient(t *testing.T, ctx context.Context, name string, derpURL st
 // received on dstCh.
 //
 // If the packet doesn't arrive within 500ms, it will try to send it again until
-// testutil.WaitLong is reached.
+// the context expires.
 //
 //nolint:revive
 func testDERPSend(t *testing.T, ctx context.Context, dstKey key.NodePublic, dstCh <-chan derp.ReceivedPacket, src *derphttp.Client) {
@@ -1150,11 +1151,17 @@ func testDERPSend(t *testing.T, ctx context.Context, dstKey key.NodePublic, dstC
 	for {
 		select {
 		case pkt := <-dstCh:
-			require.Equal(t, src.SelfPublicKey(), pkt.Source, "packet came from wrong source")
-			require.Equal(t, msg, pkt.Data, "packet data is wrong")
+			if pkt.Source != src.SelfPublicKey() {
+				t.Logf("packet came from wrong source: %s", pkt.Source)
+				continue
+			}
+			if !bytes.Equal(pkt.Data, msg) {
+				t.Logf("packet data is wrong: %s", pkt.Data)
+				continue
+			}
 			return
 		case <-ctx.Done():
-			t.Fatal("timed out waiting for packet")
+			t.Fatal("timed out waiting for valid packet")
 			return
 		case <-ticker.C:
 		}

